proposals/ref-struct-interfaces.md

```diff
@@ -277,6 +285,418 @@ Specifically by using the `CorGenericParamAttr.gpAllowByRefLike(0x0020)` or `Sys
+    {
+        using (new S2())
+        {
+        } // S2.System.IDisposable.Dispose is called
+    }
+}
+```
+
+Note that preference is given to a `Dispose` method that implements the pattern, and only if one is not found, `IDisposable`
+implementation is used.
+
+A `using` statement will recognize and use implementation of `IDisposable` interface when resource is a type parameter that 
+`allows ref struct` and `IDisposable` is in its effective interfaces set.
+```csharp
+class C
+{
+    static void Test<T>(T t) where T : System.IDisposable, allows ref struct
+    {
+        using (t)
+        {
+        }
+    }
+}
+```
+
+Note that a pattern `Dispose` method will not be recognized on a type parameter that `allows ref struct` because
+an interface (and this is the only place where we could possibly look for a pattern) is not a ref struct.
+```csharp
+interface IMyDisposable
+{
+    void Dispose();
+}
+class C
+{
+    static void Test<T>(T t, IMyDisposable s) where T : IMyDisposable, allows ref struct
+    {
+        using (t) // Error, the pattern is not recognized
+        {
+        }
+
+        using (s) // Error, the pattern is not recognized
+        {
+        }
+    }
+}
+```
+
+### `await using` statement
+
+Currently language disallows using ref structs as resources in `await using` statement. The same limitation will be
+applied to a type parameter that `allows ref struct`.
+
+There is a proposal to lift general restrictions around usage of ref structs in async methods - https://github.com/dotnet/csharplang/pull/7994.
+The remainder of the section describes behavior after the general limitation for `await using` statement will be lifted, if/when that will happen. 
+
+An `await using` statement will recognize and use implementation of `IAsyncDisposable` interface when resource is a ref struct.
+```csharp
+ref struct S2 : IAsyncDisposable
+{
+    ValueTask IAsyncDisposable.DisposeAsync()
+    {
+    }
+}
+
+class C
+{
+    static async Task Main()
+    {
+        await using (new S2())
+        {
+        } // S2.IAsyncDisposable.DisposeAsync
+    }
+}
+```
+
+Note that preference is given to a `DisposeAsync` method that implements the pattern, and only if one is not found, `IAsyncDisposable`
+implementation is used.
+
+A pattern `DisposeAsync` method will be recognized on a type parameter that `allows ref struct` as it is recognized on
+type parameters without that constraint today.
+
+```csharp
+interface IMyAsyncDisposable
+{
+    ValueTask DisposeAsync();
+}
+
+class C
+{
+    static async Task Test<T>() where T : IMyAsyncDisposable, new(), allows ref struct
+    {
+        await using (new T())
+        {
+        } // IMyAsyncDisposable.DisposeAsync
+    }
+}
+```
+
+A `using` statement will recognize and use implementation of `IAsyncDisposable` interface when resource is a type parameter that 
+`allows ref strict`, there is no `DisposeAsync` pattern match, and `IAsyncDisposable` is in type parameter's effective interfaces set.
```

Comment:
```suggestion
`allows ref struct`, there is no `DisposeAsync` pattern match, and `IAsyncDisposable` is in type parameter's effective interfaces set.
```

Reply:
> A `using` statement will recognize and use implementation of `IAsyncDisposable` interface when resource is a type parameter that 
> `allows ref strict`, there is no `DisposeAsync` pattern match, and `IAsyncDisposable` is in type parameter's effective interfaces set.

Currently, it looks like we use `IAsyncDisposable` even if there is a `DisposeAsync` pattern (available through an interface) - see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwKwG5nICWAdgC4CmATgGYCGAxuegJICyAngIIDO7x9AEULcADgHtutAEYAbcoiQBvZOlXoAarRkBXclABs6IaInkefegAoAlPiQBfAihjoAwsmVI1mOIagAOTEMAFXJuUgAeYIA+G3QAdwALKiZg9BAWDnN+Y3FJWXkAGhZswWE86Tli4nJ4mxU1T29vKABOTAxLGvj0YJtrBub0JqHmrAA6VzFibjE5cYB1SkIKSzgYAGZbQeb7dAB6fZLeHPKJSvJx3NNSnccHIA==).

Reply:
> > A `using` statement will recognize and use implementation of `IAsyncDisposable` interface when resource is a type parameter that
> > `allows ref strict`, there is no `DisposeAsync` pattern match, and `IAsyncDisposable` is in type parameter's effective interfaces set.
> 
> Currently, it looks like we use `IAsyncDisposable` even if there is a `DisposeAsync` pattern (available through an interface) - see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwKwG5nICWAdgC4CmATgGYCGAxuegJICyAngIIDO7x9AEULcADgHtutAEYAbcoiQBvZOlXoAarRkBXclABs6IaInkefegAoAlPiQBfAihjoAwsmVI1mOIagAOTEMAFXJuUgAeYIA+G3QAdwALKiZg9BAWDnN+Y3FJWXkAGhZswWE86Tli4nJ4mxU1T29vKABOTAxLGvj0YJtrBub0JqHmrAA6VzFibjE5cYB1SkIKSzgYAGZbQeb7dAB6fZLeHPKJSvJx3NNSnccHIA==).

I can see how this could be confusing, but there is no pattern match in this case due to an ambiguity.

Reply:
My comment here (and below) was about the text rather than the example. The text states we'll use the interface implementation when there is no pattern implementation, which seems to imply that if there was a pattern implementation, we'd use that instead of the interface implementation. But currently it looks like we use the interface implementation regardless of whether there is a pattern implementation.

Reply:
> there is no `DisposeAsync` pattern match

The statement doesn't talk about " a pattern implementation" there, it talks about "a pattern match", which doesn't have to exists (and in fact doesn't in some scenarios) when there is a pattern implementation.